### PR TITLE
schemachanger,*: introduce new schema changer

### PR DESCRIPTION
**Pull request by ajwerner**
This PR introduces the new schema changer, for which a partially written RFC (including a "guide-level explanation") can be found in #59288. It contains a somewhat complete implementation for a subset of `ALTER TABLE ... ADD COLUMN`. Like the existing implementation, these schema changes are planned and partially executed in the user transaction, but most of the work happens in an asynchronous job that runs after the transaction has committed. We use the index backfiller to build new columns, as described in #47989.

See the RFC and individual commits for details.

Missing things that will be in later PRs:

- Making execution changes to support concurrent writes for columns added using the index backfiller. We currently don't have a way to write column values which are not part of the primary index. See #59149. (#58990 takes care of the backfill part).
- Ensuring compatible coexistence with the old schema changer; in particular, polling for queued mutations to finish before any schema changes can start, using child transactions within the user transaction (#56588). The new schema changer assumes it has exclusive rights to perform schema changes on the table, which anticipates the transactional schema changes of the future.
- Reverting schema changes when failed or canceled.
- Properly checkpointing and reading job progress for the backfill. Currently there's a placeholder implementation that doesn't interact with the job at all.

### CODE REVIEW

**Comment by ajwerner** [comment]
Update: I pushed a revision which reworks the way that the schema changer connects to the `connExecutor`. It also reworks the setting to enable the schema changer into something more like we'd ship.

**Review by postamar — CHANGES_REQUESTED** [review]
I looked at this changeset in detail, that being said my review is necessarily superficial and focusing mostly on the implementation outlined in the RFC rather than the design because I'm not very knowledgeable about the existing logic replaced by this new schema changer. I suspect that this will be the case (to varying degrees) for anyone but @lucy-zhang and @ajwerner and that's fine; anyway the RFC itself has already been the subject of plenty of rich discussion.

I found the new code easy to follow and nicely modularized. As a consequence it allows for some powerful tests like for instance the logic tests for scbuild, which I really like. I feel this is worth merging as it is and I hope this actually gets used soon in some meaningful way to uncover bugs and so that it doesn't suffer from bitrot.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-:-MSYG5k78l9R0szVN5aB:b1bu8dh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @lucy-zhang)

___
*[pkg/sql/schema_change_plan_node.go, line 46 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYFdXr4CIowFWVAzyF:-MSYFdXr4CIowFWVAzyG:b-rgqoho) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/schema_change_plan_node.go#L46)):*
> ```Go
> 
> // schemaChangePlanNode is the planNode utilized by the new schema changer to
> //
> ```

nit: missing comment lines.
___
*[pkg/sql/schemachanger/scpb/scpb.proto, line 62 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYFS3N0RdR3ptN22lN:-MSYFS3N0RdR3ptN22lO:b2vop27) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/schemachanger/scpb/scpb.proto#L62)):*
> ```protobuf
>   uint32 table_id = 1 [(gogoproto.customname) = "TableID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.ID"];
>   cockroach.sql.sqlbase.IndexDescriptor index = 2 [(gogoproto.nullable) = false];
>   uint32 other_primary_index = 3 [(gogoproto.customname) = "OtherPrimaryIndex", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.IndexID"];
> ```

nit: shouldn't this read `other_primary_index_id` out of consistency? The same question applies to `replacement_for` below.
___
*[pkg/sql/schemachanger/scplan/rules.go, line 60 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYMiPQ02UwM1JMNW3N:-MSYMiPQ02UwM1JMNW3O:b-zebfhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/schemachanger/scplan/rules.go#L60)):*
> ```Go
> }
> 
> var rules = map[scpb.Element]targetRules{
> ```

I'm worried that laying these rules out as a map like this is going to age poorly. The literal is already one huge block with many nesting levels and with time it's only going to grow. I'm finding that it's already difficult to read and a lot of this stuff is highly sensitive, so the potential for dumb mistakes due to confusion seems really high to me. Consider what the diff of a future patch is going to look like from the point of view of the reviewer in github or reviewable: it's going to be really hard to tell whether the lines being changed are for `PrimaryIndex` or for `SecondaryIndex`, for example.

As a result I feel it's worth breaking this down into one block per key at the top level, at least. This would mitigate the potential for confusion a great deal because review tools highlight the line where a block begins (in a function definition this usually is the function name & signature). Going further, at the top level the keys of this `rules` map are used to dispatch on `Element` subtypes using reflection; perhaps this could instead be implemented with interfaces somehow?

In `forward` and `backward` I like that you list the states from `ABSENT` to `VALIDATED` in forward and backwards order respectively. Perhaps this could also be enforced in `declarative.go`? I wouldn't be shy about being overly generous with such "compile-time" checks. Speaking of which I'm surprised that you rely on panics there instead of having a `rules_test.go` of sorts that tests the validity of the rules map.

Anyway tl;dr readibility here is worth investing into some more IMHO.
___
*[pkg/sql/sessiondata/session_data.go, line 409 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYFpWO0LIZV8Mfz5RQ:-MSYFpWO0LIZV8Mfz5RR:bx0042d) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/sessiondata/session_data.go#L409)):*
> ```Go
> 
> // NewSchemaChangerMode controls if and when the new schema changer is in use.
> // Since 2.1, we run everything through the DistSQL infrastructure,
> ```

nit: did you mean 20.1 instead of 2.1?

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58979)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @postamar)

___
*[pkg/sql/schema_change_plan_node.go, line 46 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYFdXr4CIowFWVAzyF:-MSZmbAf37_W0_z_8vUJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/schema_change_plan_node.go#L46)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: missing comment lines.
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scpb/scpb.proto, line 62 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYFS3N0RdR3ptN22lN:-MSZoK744Wnu1IODc7dq:b-e3u4dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/schemachanger/scpb/scpb.proto#L62)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: shouldn't this read `other_primary_index_id` out of consistency? The same question applies to `replacement_for` below.
</blockquote></details>

Yes. Also, `replacement_for` shouldn't exist. Fixed.
___
*[pkg/sql/schemachanger/scplan/rules.go, line 60 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYMiPQ02UwM1JMNW3N:-MSZpSfB2Q_OmPTKsP9S:bozs6ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/schemachanger/scplan/rules.go#L60)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm worried that laying these rules out as a map like this is going to age poorly. The literal is already one huge block with many nesting levels and with time it's only going to grow. I'm finding that it's already difficult to read and a lot of this stuff is highly sensitive, so the potential for dumb mistakes due to confusion seems really high to me. Consider what the diff of a future patch is going to look like from the point of view of the reviewer in github or reviewable: it's going to be really hard to tell whether the lines being changed are for `PrimaryIndex` or for `SecondaryIndex`, for example.

As a result I feel it's worth breaking this down into one block per key at the top level, at least. This would mitigate the potential for confusion a great deal because review tools highlight the line where a block begins (in a function definition this usually is the function name & signature). Going further, at the top level the keys of this `rules` map are used to dispatch on `Element` subtypes using reflection; perhaps this could instead be implemented with interfaces somehow?

In `forward` and `backward` I like that you list the states from `ABSENT` to `VALIDATED` in forward and backwards order respectively. Perhaps this could also be enforced in `declarative.go`? I wouldn't be shy about being overly generous with such "compile-time" checks. Speaking of which I'm surprised that you rely on panics there instead of having a `rules_test.go` of sorts that tests the validity of the rules map.

Anyway tl;dr readibility here is worth investing into some more IMHO.
</blockquote></details>

I agree that this could use some improvements in readability. The route @ajwerner and I were thinking of going down was to have some textual declarative specification of these rules and generate code based on that, maybe resembling the optgen stuff (or at least sharing some infrastructure). But that doesn't address your point about review tools knowing how to parse code and in fact makes it worse.

I'm going to leave this commit unchanged because I suspect the route forward will become clearer once we use this part of the code more.
___
*[pkg/sql/sessiondata/session_data.go, line 409 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/58979#-MSYFpWO0LIZV8Mfz5RQ:-MSZmh-r2yLXVO8x0jjI:b-qwg8nq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ace9f91029901a58d4ee4c1985ac1517af86cf2/pkg/sql/sessiondata/session_data.go#L409)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: did you mean 20.1 instead of 2.1?
</blockquote></details>

This is copy-paste cruft, removed.

**Comment by lucy-zhang** [comment]
CI is green so I'm going to bors this now. For anyone watching, @ajwerner and I have been working on this together for a while and reviewing each other's code, and we feel confident enough to merge this with the setting off by default.

bors r+
